src/core/Akka.Remote/Transport/Helios/HeliosTcpTransport.cs

```diff
@@ -7,155 +7,106 @@
+                NotifyListener(new InboundPayload(bytes));
             }
-        }
 
-        /// <summary>
-        /// Fires whenever a Helios <see cref="IConnection"/> experienced a non-fatal error.
-        /// 
-        /// <remarks>The connection should still be open even after this event fires.</remarks>
-        /// </summary>
-        /// <param name="ex">The execption that triggered this event.</param>
-        /// <param name="erroredChannel">The handle to the Helios channel that errored.</param>
-        protected override void OnException(Exception ex, IConnection erroredChannel)
-        {
-            // Want to notify only for encoding exceptions
-            if(!(ex is HeliosConnectionException))
-                ChannelLocalActor.Notify(erroredChannel, new UnderlyingTransportError(ex, "Non-fatal network error occurred inside underlying transport"));
+            // decrease the reference count to 0 (releases buffer)
+            ReferenceCountUtil.SafeRelease(message);
         }
 
-        public override void Dispose()
+        public override void ExceptionCaught(IChannelHandlerContext context, Exception exception)
         {
-
-            ChannelLocalActor.Remove(UnderlyingConnection);
-            base.Dispose();
+            base.ExceptionCaught(context, exception);
+            NotifyListener(new Disassociated(DisassociateInfo.Unknown));
+            context.CloseAsync(); // close the channel
         }
     }
 
     /// <summary>
     /// TCP handlers for inbound connections
     /// </summary>
-    class TcpServerHandler : TcpHandlers
+    internal sealed class TcpServerHandler : TcpHandlers
     {
-        private Task<IAssociationEventListener> _associationListenerTask;
+        private readonly Task<IAssociationEventListener> _associationEventListener;
 
-        public TcpServerHandler(HeliosTransport wrappedTransport, Task<IAssociationEventListener> associationListenerTask, IConnection underlyingConnection)
-            : base(underlyingConnection)
+        public TcpServerHandler(HeliosTransport wrappedTransport, ILoggingAdapter log, Task<IAssociationEventListener> associationEventListener) : base(wrappedTransport, log)
         {
-            WrappedTransport = wrappedTransport;
-            _associationListenerTask = associationListenerTask;
+            _associationEventListener = associationEventListener;
         }
 
-        protected void InitInbound(IConnection connection, INode remoteSocketAddress, NetworkData msg)
+        public override void ChannelActive(IChannelHandlerContext context)
         {
-            _associationListenerTask.ContinueWith(r =>
+            InitInbound(context.Channel, (IPEndPoint)context.Channel.RemoteAddress, null);
+            base.ChannelActive(context);
+        }
+
+        void InitInbound(IChannel channel, IPEndPoint socketAddress, object msg)
+        {
+            // disable automatic reads
+            channel.Configuration.AutoRead = false;
```

Comment:
`AutoRead` is false by default - otherwise one of the handshake messages could make it through without an actor listening. When `AutoRead` is set back to true the channel will begin automatically reading messages from the socket as they arrive using `SocketAsyncEventArgs` internally.